### PR TITLE
[UNR-5447] Replace Handover with ServerOnly replication condition

### PR DESCRIPTION
**Pull request by mironec**
#### Description
Remove the `Handover` variable specifier and replace it just with the generic Unreal `Replication` modifier, but using a new replication condition, namely `COND_ServerOnly`, to serve the same purpose as the previous `Handover` one.

This reduces code duplication in our codebase, and renames Handover to the more appropriate ServerOnly replication, as that was the implementation.

#### Release note
The `Handover` variable specifier has been deprecated. It should be replaced with the standard `Replicated` variable specifier and restricting the replication with the new `COND_ServerOnly` replication condition in `GetLifetimeReplicatedProps`. `Handover` variables will try to replicate using the new replication condition, but support will be removed in the next release.

#### Tests
Ran all tests, quite a few times. Also ran slow tests, and the NFRs, the results are on the testgym PR.

#### Documentation
Discussed, but yes, will probably need a documentation update.

#### Reminders (IMPORTANT)
Incremented GDK version, engine version, schema db version.

#### Primary reviewers
@m-samiec 

Associated PRs:
https://github.com/spatialos/UnrealGDKTestGyms/pull/387
https://github.com/improbableio/UnrealEngine/pull/933
https://github.com/improbableio/UnrealEngine/pull/934
https://github.com/spatialos/UnrealGDKExampleProject/pull/273

### CODE REVIEW

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp [hunk 1]
```diff
@@ -406,16 +403,6 @@ void USpatialActorChannel::UpdateShadowData()
 		ResetShadowData(*ComponentReplicator.RepLayout, ComponentReplicator.ChangelistMgr->GetRepChangelistState()->StaticBuffer,
 						ActorComponent);
 	}
-
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp:409]: I can already tell this is going to be a great PR

File: SpatialGDK/Source/SpatialGDK/Private/Interop/ActorSystem.cpp [hunk 2]
```diff
@@ -666,17 +665,14 @@ void ActorSystem::ComponentUpdated(const Worker_EntityId EntityId, const Worker_
 
 	ESchemaComponentType Category = NetDriver->ClassInfoManager->GetCategoryByComponentId(ComponentId);
 
-	if (Category == SCHEMA_Data || Category == SCHEMA_OwnerOnly || Category == SCHEMA_InitialOnly)
+	if (Category != SCHEMA_Invalid)
 	{
+		if (Category == SCHEMA_ServerOnly)
+		{
+			ensureAlways(NetDriver->IsServer());
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Private/Interop/ActorSystem.cpp:672]: Prefer `ensureAlways(Category != ServerOnly || NetDriver->IsServer())`

File: SpatialGDK/Source/SpatialGDK/Private/Interop/ActorSystem.cpp [hunk 3]
```diff
@@ -1120,9 +1107,7 @@ void ActorSystem::ResolveObjectReferences(FRepLayout& RepLayout, UObject* Replic
 			continue;
 		}
 
-		// ParentIndex is -1 for handover properties
-		bool bIsHandover = ObjectReferences.ParentIndex == -1;
-		FRepParentCmd* Parent = ObjectReferences.ParentIndex >= 0 ? &RepLayout.Parents[ObjectReferences.ParentIndex] : nullptr;
+		FRepParentCmd& Parent = RepLayout.Parents[ObjectReferences.ParentIndex];
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Private/Interop/ActorSystem.cpp:1110]: nice - but make it const?

File: SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialClassInfoManager.cpp [hunk 4]
```diff
@@ -200,25 +200,10 @@ void USpatialClassInfoManager::CreateClassInfoForClass(UClass* Class)
 		}
 	}
 
-	const bool bTrackHandoverProperties = ShouldTrackHandoverProperties();
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialClassInfoManager.cpp:203]: Still reading - but previously we could prevent handover properties from replicating via this check. I assume that's not possible anymore in the current setup?

Comment (m-samiec) [reply SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialClassInfoManager.cpp:203]: ie. stop server_only properties replicating when they aren't needed

File: SpatialGDK/Source/SpatialGDKEditor/Private/SchemaGenerator/SchemaGenerator.cpp [hunk 5]
```diff
@@ -23,21 +23,23 @@ namespace
...
+	case REP_InitialOnly:
 		return SCHEMA_InitialOnly;
-	}
-	else
-	{
+	case REP_ServerOnly:
+		return SCHEMA_ServerOnly;
+	default:
 		checkNoEntry();
+		static_assert(
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDKEditor/Private/SchemaGenerator/SchemaGenerator.cpp:38]: Double static_assert here

File: SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialClassInfoManager.cpp (hunk 4 above)
Comment (mironec) [reply SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialClassInfoManager.cpp:203]: yeah, thought about this a bit, we could reimplement this, but wasn't sure how much it's needed.

File: SpatialGDK/Source/SpatialGDKEditor/Private/SchemaGenerator/SchemaGenerator.cpp (hunk 5 above)
Comment (m-samiec) [reply SpatialGDK/Source/SpatialGDKEditor/Private/SchemaGenerator/SchemaGenerator.cpp:38]: Oh they're asserting on different values

File: SpatialGDK/Source/SpatialGDKEditor/Private/SchemaGenerator/SchemaGenerator.cpp [hunk 6]
```diff
@@ -569,6 +481,29 @@ void GenerateSubobjectSchema(FComponentIdGenerator& IdGenerator, UClass* Class,
 	SubobjectClassPathToSchema.Add(Class->GetPathName(), SubobjectSchemaData);
 }
 
+EReplicatedPropertyGroup SchemaComponentTypeToPropertyGroup(ESchemaComponentType SchemaType)
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDKEditor/Private/SchemaGenerator/SchemaGenerator.cpp:484]: nit - would rather this was after PropertyGroupToSchemaComponentType

File: SpatialGDK/Source/SpatialGDKEditor/Private/SchemaGenerator/SpatialGDKEditorSchemaGenerator.cpp [hunk 7]
```diff
@@ -556,11 +529,13 @@ FString GetComponentSetNameBySchemaType(ESchemaComponentType SchemaType)
 		return SpatialConstants::DATA_COMPONENT_SET_NAME;
 	case SCHEMA_OwnerOnly:
 		return SpatialConstants::OWNER_ONLY_COMPONENT_SET_NAME;
-	case SCHEMA_Handover:
+	case SCHEMA_ServerOnly:
 		return SpatialConstants::HANDOVER_COMPONENT_SET_NAME;
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDKEditor/Private/SchemaGenerator/SpatialGDKEditorSchemaGenerator.cpp:533]: HANDOVER_COMPONENT_SET_NAME

File: SpatialGDK/Source/SpatialGDKEditor/Private/SchemaGenerator/TypeStructure.cpp [hunk 8]
```diff
@@ -12,21 +12,30 @@ using namespace SpatialGDKEditor::Schema;
...
-	{
+	case REP_InitialOnly:
 		return TEXT("InitialOnly");
-	}
-	else
-	{
+	case REP_ServerOnly:
+		return TEXT("ServerOnly");
+	default:
+		static_assert(REP_Count == 4, "Unexpected number of ReplicatedPropertyGroups, please update this function.");
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDKEditor/Private/SchemaGenerator/TypeStructure.cpp:38]: nit - here and elsewhere - I'd put these static asserts prior to the beginning of the switch statement

File: SpatialGDK/Source/SpatialGDKEditor/Private/SchemaGenerator/SpatialGDKEditorSchemaGenerator.cpp [hunk 9]
```diff
@@ -659,23 +636,8 @@ void WriteServerAuthorityComponentSet(const USchemaDatabase* SchemaDatabase, con
...
-					case SCHEMA_Handover:
-						Writer.Printf("unreal.generated.{0}.{1}Handover,", ActorClassName.ToLower(), ActorClassName);
-						break;
-					case SCHEMA_InitialOnly:
-						Writer.Printf("unreal.generated.{0}.{1}InitialOnly,", ActorClassName.ToLower(), ActorClassName);
-						break;
-					default:
-						break;
-					}
+					Writer.Printf("unreal.generated.{0}.{1}{2},", ActorClassName.ToLower(), ActorClassName,
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDKEditor/Private/SchemaGenerator/SpatialGDKEditorSchemaGenerator.cpp:639]: nice

File: SpatialGDK/Source/SpatialGDKEditor/Private/SchemaGenerator/SchemaGenerator.cpp (hunk 5 above)
Comment (mironec) [reply SpatialGDK/Source/SpatialGDKEditor/Private/SchemaGenerator/SchemaGenerator.cpp:38]: This is intentional, one for REPs and one for SCHEMAs, but in general I may have gone a bit overboard with the static_asserts.

File: SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialClassInfoManager.cpp [hunk 10]
```diff
@@ -353,20 +307,6 @@ void USpatialClassInfoManager::FinishConstructingSubobjectClassInfo(const FStrin
...
-	// There's currently a bug that lets handover data get sent to clients in the initial
-	// burst of data for an entity, which leads to log spam in the SpatialReceiver. By tracking handover
-	// properties on clients, we can prevent that spam. Cannot be removed yet because of Kraken,
-	// UNR-4358 will remove this in a squid-only world.
-	if (!NetDriver->IsServer())
-	{
-		return true;
-	}
-
-	return USpatialStatics::IsHandoverEnabled(NetDriver);
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialClassInfoManager.cpp:367]: `IsHandoverEnabled` seems to make little sense now - we should discuss

**Comment by m-samiec** [comment]
Increment engine/gdk versions

File: SpatialGDK/Source/SpatialGDKEditor/Private/SchemaGenerator/SchemaGenerator.cpp (hunk 6 above)
Comment (mironec) [reply SpatialGDK/Source/SpatialGDKEditor/Private/SchemaGenerator/SchemaGenerator.cpp:484]: Ah, yes, I also wanted it like that, unfortunately PropertyGroupToSchemaComponentType is in an anonymous namespace, and I have to export this one :(
Maybe something else could be done?

**Comment by razvan-at-improbable** [comment]
Should this property just be called "Replicated" and not something more descriptive?

File: SpatialGDK/Source/SpatialGDKEditor/Private/SchemaGenerator/SpatialGDKEditorSchemaGenerator.cpp (hunk 7 above)
Comment (mironec) [reply SpatialGDK/Source/SpatialGDKEditor/Private/SchemaGenerator/SpatialGDKEditorSchemaGenerator.cpp]: Changed to SERVERONLY_COMPONENT_SET_NAME now

File: CHANGELOG.md [hunk 11]
```diff
@@ -11,6 +11,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 ## [`x.y.z`] - Unreleased
 
 ### Breaking changes:
+- The `Handover` variable specifier has been deprecated. It should be replaced with just the normal `Replicated` variable specifier and restricting the replication with the new `COND_ServerOnly` replication condition in `GetLifetimeReplicatedProps`. `Handover` variables will try to replicate using the new replication condition, but we recommend converting variables manually as soon as possible.
```
Comment (m-samiec) [inline CHANGELOG.md:14]:
```suggestion
- The `Handover` variable specifier has been deprecated. It should be replaced with the standard `Replicated` variable specifier and restricting the replication with the new `COND_ServerOnly` replication condition in `GetLifetimeReplicatedProps`. `Handover` variables will try to replicate using the new replication condition, but support will be removed in the next release.
```

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp [hunk 12]
```diff
@@ -461,17 +448,6 @@ FRepChangeState USpatialActorChannel::CreateInitialRepChangeState(TWeakObjectPtr
 	return { InitialRepChanged, *Replicator.RepLayout };
 }
 
-FHandoverChangeState USpatialActorChannel::CreateInitialHandoverChangeState(const FClassInfo& ClassInfo)
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp:464]: FHandoverChangeState can be removed from `RepDataUtils.h` now also

File: SpatialGDK/Source/SpatialGDKEditor/Private/SchemaGenerator/SchemaGenerator.cpp (hunk 6 above)
Comment (m-samiec) [reply SpatialGDK/Source/SpatialGDKEditor/Private/SchemaGenerator/SchemaGenerator.cpp:484]: er.. you could move this to namespace, and expose a function that just calls into it?
not fussed though

Comment (m-samiec) [reply SpatialGDK/Source/SpatialGDKEditor/Private/SchemaGenerator/SchemaGenerator.cpp:484]: perhaps we should expose PropertyGroupToSchemaComponentType ?

File: CHANGELOG.md [hunk 13]
```diff
@@ -11,6 +11,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 ## [`x.y.z`] - Unreleased
 
 ### Breaking changes:
+- The `Handover` variable specifier has been deprecated. It should be replaced with the standard `Replicated` variable specifier and restricting the replication with the new `COND_ServerOnly` replication condition in `GetLifetimeReplicatedProps`. `Handover` variables will try to replicate using the new replication condition, but support will be removed in the next release.
```
Comment (m-samiec) [inline CHANGELOG.md:14]: The LBStrategy still has RequiresHandoverData, not used anymore but might be useful in future so I agree with not removing. Perhaps rename to RequiresServerToServerData / RequiresServerData though ? If so would need a breaking change here also.

**Review by m-samiec — APPROVED** [review]
Amazing - great work!

**Pull request merged**